diff --git a/libusb/os/linux_usbfs.c b/libusb/os/linux_usbfs.c
index 285d9ca..0c3aa96 100644
--- a/libusb/os/linux_usbfs.c
+++ b/libusb/os/linux_usbfs.c
@@ -313,7 +313,16 @@ static int get_kernel_version(struct libusb_context *ctx,
 		usbi_err(ctx, "uname failed, errno=%d", errno);
 		return -1;
 	}
+#ifdef __OHOS__
+	if (strcmp(uts.sysname, "Linux") != 0) {
+		ver->major = 5;
+		ver->minor = 10;
+		ver->sublevel = 0;
 
+		usbi_dbg(ctx, "reported kernel version as 5.10.0");
+		return 0;
+	}
+#endif
 	atoms = sscanf(uts.release, "%d.%d.%d", &ver->major, &ver->minor, &ver->sublevel);
 	if (atoms < 2) {
 		usbi_err(ctx, "failed to parse uname release '%s'", uts.release);
