diff --git a/cups/ppd-cache.c b/cups/ppd-cache.c
index 786c66b..65aa200 100644
--- a/cups/ppd-cache.c
+++ b/cups/ppd-cache.c
@@ -1746,7 +1746,8 @@ _ppdCacheCreateWithPPD(ppd_file_t *ppd)	/* I - PPD file */
 
   if ((ppd_attr = ppdFindAttr(ppd, "cupsSingleFile", NULL)) != NULL)
     pc->single_file = !_cups_strcasecmp(ppd_attr->value, "true");
-
+  else
+    pc->single_file = 1;
  /*
   * Copy the product string, if any...
   */
